File: searchcc/sparta.cc
// Copyright (c) Facebook, Inc. and its affiliates.
// All rights reserved.
//
// This source code is licensed under the license found in the
// LICENSE file in the root directory of this source tree.
//
#include <chrono>
#include <future>

#include "searchcc/sparta.h"

namespace search {

float searchMove(
    const hle::HanabiState& state,
    hle::HanabiMove move,
    const std::vector<std::vector<hle::HanabiCardValue>>& hands,
    const std::vector<int> seeds,
    int myIdx,
    const std::vector<HybridModel>& players) {
  std::vector<std::vector<HybridModel>> allPlayers(hands.size(), players);
  std::vector<GameSimulator> games;
  games.reserve(hands.size());
  for (size_t i = 0; i < hands.size(); ++i) {
    std::vector<SimHand> simHands{
        SimHand(myIdx, hands[i]),
    };
    games.emplace_back(state, simHands, seeds[i]);
  }

  size_t terminated = 0;
  std::vector<int> notTerminated;
  for (size_t i = 0; i < games.size(); ++i) {
    assert(!games[i].terminal());
    games[i].step(move);
    if (!games[i].terminal()) {
      notTerminated.push_back(i);
    } else {
      ++terminated;
    }
  }

  while (!notTerminated.empty()) {
    std::vector<int> newNotTerminated;
    for (auto i : notTerminated) {
      assert(!games[i].state().IsTerminal());
      for (auto& actor : allPlayers[i]) {
        actor.observeBeforeAct(games[i], 0);
      }
    }

    for (auto i : notTerminated) {
      auto& game = games[i];
      int action = -1;
      for (auto& actor : allPlayers[i]) {
        int a = actor.decideAction(game, false);
        if (actor.index == game.state().CurPlayer()) {
          action = a;
        }
      }
      // std::cout << "move: " << game.getMove(action).ToString() << std::endl;
      game.step(game.getMove(action));
      if (!game.terminal()) {
        newNotTerminated.push_back(i);
      } else {
        ++terminated;
      }
    }

    notTerminated = newNotTerminated;
  }
  assert(terminated == games.size());

  std::vector<float> scores(games.size());
  float mean = 0;
  for (size_t i = 0; i < games.size(); ++i) {
    assert(games[i].terminal());
    scores[i] = games[i].score();
    mean += scores[i];
  }
  mean = mean / scores.size();
  return mean;
}

// should be called after decideAction?
hle::HanabiMove SpartaActor::spartaSearch(
    const GameSimulator& env, hle::HanabiMove bpMove, int numSearch, float threshold) {
  torch::NoGradGuard ng;

  const auto& state = env.state();
  assert(state.CurPlayer() == index);
  auto legalMoves = state.LegalMoves(index);

  if (legalMoves.empty()) {
    return bpMove;
  }

  int numSearchPerMove = numSearch / legalMoves.size();
  std::cout << "SPARTA will run " << numSearchPerMove << " searches on "
            << legalMoves.size() << " legal moves" << std::endl;
  auto simHands = handDist_.sampleHands(numSearchPerMove, &rng_);
  std::vector<int> seeds;
  for (size_t i = 0; i < simHands.size(); ++i) {
    seeds.push_back(int(rng_()));
  }

  std::vector<std::future<float>> futMoveScores;
  std::vector<HybridModel> players;
  for (auto& p : partners_) {
    players.push_back(p->model_);
  }
  for (auto move : legalMoves) {
    futMoveScores.push_back(std::async(
        std::launch::async, searchMove, state, move, simHands, seeds, index, players));
  }

  hle::HanabiMove bestMove = bpMove;
  float bpScore = -1;
  float bestScore = -1;

  std::cout << "SPARTA scores for moves:" << std::endl;
  for (size_t i = 0; i < legalMoves.size(); ++i) {
    float score = futMoveScores[i].get();
    auto move = legalMoves[i];
    if (move == bpMove) {
      assert(bpScore == -1);
      bpScore = score;
    }
    if (score > bestScore) {
      bestScore = score;
      bestMove = move;
    }
    std::cout << move.ToString() << ": " << score << std::endl;
  }

  std::cout << "SPARTA best - bp: " << bestScore - bpScore << std::endl;
  if (bestScore - bpScore >= threshold) {
    std::cout << "SPARTA changes move from " << bpMove.ToString() << " to "
              << bestMove.ToString() << std::endl;
    return bestMove;
  } else {
    return bpMove;
  }
}

}  // namespace search

File: pyhanabi/sparta.py
# Copyright (c) Facebook, Inc. and its affiliates.
# All rights reserved.
#
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.
#
import os
import sys
import argparse
import pprint
import torch
import numpy as np

# c++ backend
import set_path

set_path.append_sys_path()
import rela
import hanalearn
import utils
import common_utils


def run(seed, actors, search_actor_idx, num_search, threshold, num_thread):
    params = {
        "players": str(len(actors)),
        "seed": str(seed),
        "bomb": str(0),
        "hand_size": str(5),
        "random_start_player": str(0),  # do not randomize start_player
    }
    game = hanalearn.GameSimulator(params)
    step = 0
    moves = []
    while not game.terminal():
        print("================STEP %d================" % step)
        print(game.state().to_string())

        cur_player = game.state().cur_player()

        actors[search_actor_idx].update_belief(game, num_thread)
        for i, actor in enumerate(actors):
            actor.observe(game)

        for i, actor in enumerate(actors):
            print(f"---Actor {i} decide action---")
            action = actor.decide_action(game)
            if i == cur_player:
                move = game.get_move(action)

        # run sparta, this may change the move
        if cur_player == search_actor_idx:
            move = actors[search_actor_idx].sparta_search(
                game, move, num_search, threshold
            )

        print(f"Acitve Player {cur_player} pick action: {move.to_string()}")
        moves.append(move)
        game.step(move)
        step += 1

    print(f"Final Score: {game.get_score()}, Seed: {seed}")
    return moves, game.get_score()


def parse_args():
    parser = argparse.ArgumentParser(description="")

    parser.add_argument("--save_dir", type=str, default="exps/sparta")
    parser.add_argument("--num_search", type=int, default=10000)
    parser.add_argument("--threshold", type=float, default=0.05)
    parser.add_argument("--device", type=str, default="cuda:0")
    parser.add_argument("--num_thread", type=int, default=10)
    parser.add_argument("--weight_file", type=str, default=None)
    parser.add_argument("--num_player", type=int, default=2)
    parser.add_argument("--search_player", type=int, default=1)
    parser.add_argument("--seed", type=int, default=200191)
    parser.add_argument("--game_seed", type=int, default=19)

    args = parser.parse_args()
    return args


if __name__ == "__main__":
    torch.backends.cudnn.deterministic = True
    args = parse_args()

    logger_path = os.path.join(args.save_dir, "train.log")
    sys.stdout = common_utils.Logger(logger_path)

    if "fc_v.weight" in torch.load(args.weight_file).keys():
        bp, config = utils.load_agent(args.weight_file, {"device": args.device})
        assert not config["hide_action"]
        assert not config["boltzmann_act"]
    else:
        bp = utils.load_supervised_agent(args.weight_file, args.device)
    bp.train(False)

    bp_runner = rela.BatchRunner(bp, args.device, 2000, ["act"])
    bp_runner.start()

    seed = args.seed
    actors = []
    for i in range(args.num_player):
        actor = hanalearn.SpartaActor(i, bp_runner, seed)
        seed += 1
        actors.append(actor)

    actors[args.search_player].set_partners(actors)

    moves, score = run(
        args.game_seed,
        actors,
        args.search_player,
        args.num_search,
        args.threshold,
        args.num_thread,
    )
